### PR TITLE
Feature: Add JUnit 5 Support

### PR DESCRIPTION
**Pull request by mojtab23**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -66,6 +66,34 @@ public void test() throws Exception {
...
+        // Delete table data.
+        hbaseExtension.deleteTableData("table");
+        // You expect 0 rows in HBase table after cleaning.
+        assertEquals(0, hbaseExtension.countRows("table"));
+        }
+```
+
 ### Add it to your project
 
 You can reference to this library by either of java build systems (Maven, Gradle, SBT or Leiningen) using snippets from this jitpack link:
```
Comment (saeidrastak) [inline README.md]: Also add:
JUnit 4 and 5 dependencies are marked as optional, so you need to provide JUnit 4 or 5 dependency
(based on what version you need, and you use) in you project to make it work.

File: src/main/java/ir/sahab/hbaserule/HbaseExtension.java [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+import java.net.ServerSocket;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+
+/**
+ * JUnit 5 extension which provides a mini cluster of Hbase, DFS, and ZooKeeper.
+ */
+public class HbaseExtension implements BeforeAllCallback, AfterAllCallback, BeforeEachCallback, AfterEachCallback {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/HbaseExtension.java]: Move the common parts in HbaseExtension, HbaseRule to a base class.

File: README.md [hunk 3]
```diff
@@ -66,6 +66,34 @@ public void test() throws Exception {
...
+        // You expect number of rows already put in HBase table.
+        assertEquals(10 , hbaseExtension.countRows("tableOnBuilder"));
+        // Delete table data.
+        hbaseExtension.deleteTableData("table");
+        // You expect 0 rows in HBase table after cleaning.
+        assertEquals(0, hbaseExtension.countRows("table"));
+        }
+```
+
 ### Add it to your project
```
Comment (saeidrastak) [inline README.md:114]: Change the comment of first line of readme file and explain this library provides Junit 4 Rule and Junit 5 Extension for Hbase component.

File: README.md (hunk 1 above)
Comment (saeidrastak) [inline README.md]: Also add:
JUnit 4 and 5 dependencies are marked as optional, so you need to provide JUnit 4 or 5 dependency
(based on what version you need, and you use) in you project to make it work.

File: src/main/java/ir/sahab/hbaserule/HbaseRule.java [hunk 4]
```diff
@@ -1,31 +1,24 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+
 /**
- * JUnit rule which provides a mini cluster of Hbase, DFS, and ZooKeeper.
+ * JUnit 4 rule which provides a mini cluster of Hbase, DFS, and ZooKeeper.
  */
 public class HbaseRule extends ExternalResource {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/HbaseRule.java]: public class HbaseRule extends HbaseResource implements TestRule {

File: pom.xml [hunk 5]
```diff
@@ -16,15 +16,36 @@
     </repositories>
 
     <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
```
Comment (saeidrastak) [inline pom.xml:19]: Increase version to 1.2.0

File: src/main/java/ir/sahab/hbaserule/Builder.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
+package ir.sahab.hbaserule;
+
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.util.Bytes;
+
+import java.util.Arrays;
+import java.util.Map;
+
+class Builder<T extends HbaseBase> {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/Builder.java:14]: Add java doc

File: src/main/java/ir/sahab/hbaserule/HbaseBase.java [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+import java.net.ServerSocket;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Stream;
+
+abstract class HbaseBase {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/HbaseBase.java:27]: Add java doc

File: src/main/java/ir/sahab/hbaserule/HbaseBase.java [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+    HBaseTestingUtility utility;
+    MiniZooKeeperCluster zkCluster;
+    File logDir;
+
+    List<String> nameSpaces = new ArrayList<>();
+    List<HBaseTableDef> hBaseTableDefs = new ArrayList<>();
+
+    HbaseBase() {
+    }
+
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/HbaseBase.java:39]: remove extra empty line

File: src/main/java/ir/sahab/hbaserule/HbaseBase.java [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+    static Integer anOpenPort() {
+        try (ServerSocket socket = new ServerSocket(0)) {
+            return socket.getLocalPort();
+        } catch (IOException e) {
+            throw new AssertionError("Unable to find an open port.", e);
+        }
+    }
+
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/HbaseBase.java:47]: remove extra empty line

File: src/main/java/ir/sahab/hbaserule/HbaseExtension.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    public static final TestInstance.Lifecycle lifecycle = TestInstance.Lifecycle.PER_CLASS;
+
+    private HbaseExtension() {
+    }
+
+    public static Builder<HbaseExtension> newBuilder() {
+        return new Builder<>(new HbaseExtension());
+    }
+
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/HbaseExtension.java:20]: remove extra empty line

File: src/main/java/ir/sahab/hbaserule/HbaseRule.java [hunk 11]
```diff
@@ -1,224 +1,25 @@
...
-        utility.truncateTable(tableName).close();
-    }
-
-    public void truncateTable(TableName tableName) throws IOException {
-        utility.truncateTable(tableName).close();
+    public static Builder<HbaseRule> newBuilder() {
+        return new Builder<>(new HbaseRule());
     }
 
     static Integer anOpenPort() {
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/HbaseRule.java]: move anOpenPort method to Base class.

File: src/test/java/ir/sahab/hbaserule/HbaseExtensionTest.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import java.io.IOException;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class HbaseExtensionTest {
+    private static final String NAME_SPACE = "builderNameSpace";
```
Comment (saeidrastak) [inline src/test/java/ir/sahab/hbaserule/HbaseExtensionTest.java:17]: add line break before line 20

File: src/test/java/ir/sahab/hbaserule/HbaseExtensionTest.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    static Connection createConnection(String zkAddresses, int operationTimeout) throws IOException {
+        // Create HBase client config
+        Configuration hbaseConf = HBaseConfiguration.create();
+        hbaseConf.set(HConstants.ZOOKEEPER_QUORUM, zkAddresses);
+        hbaseConf.setInt(HConstants.HBASE_RPC_TIMEOUT_KEY, operationTimeout);
+        return ConnectionFactory.createConnection(hbaseConf);
+    }
+
```
Comment (saeidrastak) [inline src/test/java/ir/sahab/hbaserule/HbaseExtensionTest.java]: Move this method to Hbase Base class also refactor Junit4 rule unit test class.

File: src/main/java/ir/sahab/hbaserule/Builder.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    public Builder<T> addTable(TableName tableName, byte[]... columnFamilies) {
+        hbase.hBaseTableDefs.add(new HbaseBase.HBaseTableDef(tableName.getName(), columnFamilies));
+        return this;
+    }
+
+    public T build() {
+        return hbase;
+    }
+
+
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/Builder.java]: Remove extra empty lines

File: src/main/java/ir/sahab/hbaserule/Builder.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.Arrays;
+import java.util.Map;
+
+/**
+ * Builder class for building {@link HbaseExtension} and {@link HbaseRule}
+ *
+ * @param <T> {@link HbaseExtension} or {@link HbaseRule}
+ */
+class Builder<T extends HbaseBase> {
+
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/Builder.java:15]: Remove extra empty line. Please check these type of rule (this is very annoying)

Comment (mojtab23) [reply src/main/java/ir/sahab/hbaserule/Builder.java:15]: I added a checkstyle rule for this in my IDE.

File: src/main/java/ir/sahab/hbaserule/Builder.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/Builder.java]: Remove extra empty line

File: src/main/java/ir/sahab/hbaserule/Builder.java (hunk 14 above)
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/Builder.java]: Remove extra empty line

File: src/main/java/ir/sahab/hbaserule/HbaseRule.java [hunk 17]
```diff
@@ -1,311 +1,51 @@
...
-
-    public Configuration getConfiguration() {
-        return utility.getConfiguration();
+    public static Builder<HbaseRule> newBuilder() {
+        return new Builder<>(new HbaseRule());
     }
 
-    public int countRows(TableName tableName) throws IOException {
-        return utility.countRows(tableName);
+    //copied from org.junit.rules.ExternalResource
```
Comment (saeidrastak) [inline src/main/java/ir/sahab/hbaserule/HbaseRule.java]: Remove this line

**Pull request merged**